core/src/main/java/org/infinispan/persistence/sifs/IndexNode.java

```diff
@@ -761,12 +777,48 @@ private IndexNode copyWith(int cacheSegment, byte[] key, int file, int offset, i
       return new IndexNode(segment, newPrefix, newKeyParts, newLeafNodes);
    }
 
+   private int getIterationPoint(byte[] key, int cacheSegment) {
+      int comp = compare(key, prefix, prefix.length);
+      int insertionPoint;
+      if (comp > 0) {
+         insertionPoint = 0;
+      } else if (comp < 0) {
+         insertionPoint = keyParts.length;
+      } else {
+         byte[] keyPostfix = substring(key, prefix.length, key.length);
+         insertionPoint = Arrays.binarySearch(keyParts, keyPostfix, (o1, o2) -> IndexNode.compare(o2, o1));
+         if (insertionPoint < 0) {
+            insertionPoint = -insertionPoint - 1;
+         } else {
+            int segmentToUse = cacheSegment < 0 ? UnsignedNumeric.readUnsignedInt(key, 0) : cacheSegment;
+            if (UnsignedNumeric.sizeUnsignedInt(segmentToUse) < key.length) {
+               // When the length is bigger than a cache segment, that means the index prefix is a specific key and if it
+               // is equal we have to skip two spaces
```

Comment:
I was trying to follow this explanation but there are just too many new terms (for me), like "cache segment" (first I thought meant the segment in the CH, but now I'm not so sure any more), "index prefix" vs. "Segment Prefix", "KeyParts".

Is there a design document with the index node layout? If yes, it would be really helpful to have a reference to it in the class javadoc.

Reply:
> I was trying to follow this explanation but there are just too many new terms (for me), like "cache segment" (first I thought meant the segment in the CH, but now I'm not so sure any more), "index prefix" vs. "Segment Prefix", "KeyParts".

Assume everything that is just "segment" as being the segment in the SIFS index. I tried whenever possible to call the cache based segments as cacheSegment, but I am sure I missed some like `segmentToUse` should be `cacheSegmentToUse`

Segment Prefix is the byte representation of the segment that is appended to every value in the index, so they are sorted by "cache segment".

Key Parts is a subset of first parts of the bytes that determine the B+ child lookup.

> Is there a design document with the index node layout? If yes, it would be really helpful to have a reference to it in the class javadoc.

The index nodes are laid out using a standard B+ structure.

I can add some more info to the class Javadoc at least.

Reply:
As always, :+1: from me to define the terms in the class javadoc and use the same terms throughout.